db/buffered_upsert.go

```diff
@@ -0,0 +1,193 @@
+package db
+
+import (
+	"context"
+	"time"
+
+	"github.com/mongodb/grip"
+	"github.com/pkg/errors"
+	"gopkg.in/mgo.v2/bson"
+)
+
+type anserBufUpsertImpl struct {
+	opts    BufferedWriteOptions
+	db      Database
+	cancel  context.CancelFunc
+	upserts chan upsertOp
+	flusher chan chan error
+	closer  chan chan error
+	err     chan error
+}
+
+type upsertOp struct {
+	query  interface{}
+	record interface{}
+}
+
+func (bu *anserBufUpsertImpl) start(ctx context.Context) {
```

Comment:
We might want to keep the struct with this logic generic and only vary the implementation for the actual writes, as most of this is similar to the code in buffered_insert.go.

Reply:
eh, I'm ok if this is duplicated, the ops are a bit different.